File: JAVA/24.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if(head == null || head.next == null) return head;

        //递归
        // ListNode one = head, two = head.next;
        // one.next = swapPairs(two.next);
        // two.next = one;
        // return two;

        //迭代
        ListNode dummy = new ListNode(0, head);
        ListNode zero = dummy, one = head;

        while(one != null && one.next != null) {
            ListNode two = one.next;
            one.next = two.next;
            two.next = one;
            zero.next = two;

            zero = one;
            one = one.next;
        }

        return dummy.next;
    }
}

File: GO/24.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	//递归
	// one := head
	// two := head.Next
	// one.Next = swapPairs(two.Next)
	// two.Next = one

	// return two

	//迭代
	dummy := &ListNode{Next: head}
	zero := dummy
	one := head

	for one != nil && one.Next != nil {
		two := one.Next
		one.Next = two.Next
		two.Next = one
		zero.Next = two

		//update
		zero = one
		one = one.Next
	}
	return dummy.Next
}